File: src/main/java/net/miarma/mkernel/commands/misc/PayXpCommand.java
package net.miarma.mkernel.commands.misc;

import java.util.List;

import org.bukkit.Bukkit;
import org.bukkit.Sound;
import org.bukkit.entity.Player;

import dev.jorel.commandapi.CommandAPICommand;
import net.miarma.mkernel.config.CommandWrapper;
import net.miarma.mkernel.config.providers.CommandProvider;
import net.miarma.mkernel.config.providers.MessageProvider;
import net.miarma.mkernel.util.MessageUtil;

public class PayXpCommand {
    public static void register() {
        CommandWrapper payXpCmd = CommandProvider.getPayXpCommand();
        new CommandAPICommand(payXpCmd.getName())
            .withArguments(CommandProvider.Arguments.playerArg(), CommandProvider.Arguments.levels())
            .withFullDescription(payXpCmd.getDescription())
            .withPermission(payXpCmd.getPermission().base())
            .withShortDescription(payXpCmd.getDescription())
            .executesPlayer((sender, args) -> {
                Player victim = Bukkit.getPlayer(args.getRaw(0));
                Integer cantidad = Integer.valueOf(args.getRaw(1));

                if(args.count() > 2) {
                    MessageUtil.sendMessage(sender, MessageProvider.Errors.tooManyArguments(), true);
                }

                if(sender.getLevel()>0) {
                    sender.setLevel(sender.getLevel()-cantidad);
                    victim.setLevel(victim.getLevel()+cantidad);
                    MessageUtil.sendMessage(sender, payXpCmd.getMessages()[0], true,
                                                 List.of("%target%", "%amount%"),
                                                 List.of(victim.getName(), cantidad.toString()));
                    MessageUtil.sendMessage(victim, payXpCmd.getMessages()[1], true,
                                                 List.of("%sender%", "%amount%"),
                                                 List.of(sender.getName(), cantidad.toString()));
                    sender.playSound(sender, Sound.BLOCK_NOTE_BLOCK_PLING, 1, 1);
                    victim.playSound(victim, Sound.ENTITY_EXPERIENCE_ORB_PICKUP, 1, 1);
                } else {
                    MessageUtil.sendMessage(sender, MessageProvider.Errors.notEnoughLevels(), true);
                }
            })
            .register();
    }
}

File: src/main/kotlin/net/miarma/mkernel/command/impl/misc/PayXpCommand.kt
package net.miarma.mkernel.command.impl.misc

import com.google.inject.Inject
import com.google.inject.Singleton
import dev.jorel.commandapi.arguments.IntegerArgument
import dev.jorel.commandapi.arguments.PlayerProfileArgument
import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.playerExecutor
import net.miarma.mkernel.command.MCommand
import net.miarma.mkernel.common.config.ConfigKeys
import net.miarma.mkernel.common.service.impl.ConfigService
import net.miarma.mkernel.common.service.impl.MessageService
import net.miarma.mkernel.util.PlayerUtil
import org.bukkit.Sound

@Singleton
class PayXpCommand @Inject constructor(private val configService: ConfigService, private val messageService: MessageService) : MCommand {
    override fun register() {
        commandAPICommand(configService.getString(ConfigKeys.Commands.PayXp.NAME)) {
            withArguments(
                PlayerProfileArgument(configService.getString(ConfigKeys.Arguments.PLAYER)),
                IntegerArgument(configService.getString(ConfigKeys.Arguments.LEVELS), 1)
            )
            withFullDescription(configService.getString(ConfigKeys.Commands.PayXp.DESC))
            withPermission(configService.getString(ConfigKeys.Commands.PayXp.PERM))
            withUsage(configService.getString(ConfigKeys.Commands.PayXp.USAGE))
            playerExecutor { sender, args ->
                val target = PlayerUtil.fromArg(args[0])
                val amount = args[1] as Int

                if (target == null) {
                    messageService.builder(configService.getString(ConfigKeys.Messages.General.Errors.PLAYER_NOT_FOUND)).withPrefix().send(sender)
                    return@playerExecutor
                }

                if (sender.level >= amount) {
                    sender.level -= amount
                    target.level += amount

                    messageService.builder(configService.getString(ConfigKeys.Commands.PayXp.MSG_YOU_OTHERS)).withPrefix().tag("target", target.name).tag("amount", amount.toString()).send(sender)
                    messageService.builder(configService.getString(ConfigKeys.Commands.PayXp.MSG_OTHERS_YOU)).withPrefix().forPlayer(target).tag("sender", sender.name).tag("amount", amount.toString()).send(target)

                    sender.playSound(sender.location, Sound.BLOCK_NOTE_BLOCK_PLING, 1.0f, 1.0f)
                    target.playSound(target.location, Sound.ENTITY_EXPERIENCE_ORB_PICKUP, 1.0f, 1.0f)
                } else {
                    messageService.builder(configService.getString(ConfigKeys.Messages.General.Errors.NOT_ENOUGH_LEVELS)).withPrefix().send(sender)
                }
            }
        }
    }
}